packages/back-end/src/controllers/datasources.ts

```diff
@@ -860,3 +873,298 @@ export async function fetchBigQueryDatasets(
+  }
+}
+
+export async function deleteMaterializedColumn(
+  req: AuthRequest<null, { datasourceId: string; matColumnName: string }>,
+  res: Response
+) {
+  const context = getContextFromReq(req);
+  const { datasourceId, matColumnName } = req.params;
+
+  const datasource = await getDataSourceById(context, datasourceId);
+  if (!datasource) {
+    throw new Error("Cannot find datasource");
+  }
+
+  if (!context.permissions.canUpdateDataSourceSettings(datasource)) {
+    context.permissions.throwPermissionError();
+  }
+
+  if (
+    !isDataSourceType<GrowthbookClickhouseDataSource>(
+      datasource,
+      "growthbook_clickhouse"
+    )
+  ) {
+    throw new Error(
+      "Can only manage materialized columns for growthbook-clickhouse datasources"
+    );
+  }
+
+  const originalColumns = datasource.settings.materializedColumns || [];
+
+  const originalIdx = originalColumns.findIndex(
+    (col) => col.columnName === matColumnName
+  );
+  if (originalIdx === -1) {
+    throw new Error(`Cannot find materialized column ${matColumnName}`);
+  }
+  const finalColumns = [
+    ...originalColumns.slice(0, originalIdx),
+    ...originalColumns.slice(originalIdx + 1),
+  ];
+
+  const updates: Pick<GrowthbookClickhouseDataSource, "settings"> = {
+    settings: {
+      ...datasource.settings,
+      materializedColumns: finalColumns,
+    },
+  };
+
+  try {
+    await updateMaterializedColumns({
+      datasource,
+      columnsToAdd: [],
+      columnsToDelete: [matColumnName],
+      columnsToRename: [],
+      finalColumns,
+      originalColumns,
+    });
+    await updateDataSource(context, datasource, updates);
+
+    const integration = getSourceIntegrationObject(context, {
+      ...datasource,
+      ...updates,
+    });
+
+    res.status(200).json({
+      status: 200,
+      datasource: getDataSourceWithParams(integration),
+    });
+  } catch (e) {
+    req.log.error(e, "Failed to update data source");
+    res.status(500).json({
+      status: 500,
+      message: e.message || "An error occurred",
+    });
+  }
+}
+
+function sanitizeMatColumnInput(userInput: MaterializedColumn) {
+  if (!factTableColumnTypes.includes(userInput.datatype)) {
+    throw new Error("Invalid datatype");
+  }
+  const sourceField = sanitizeMatColumnSourceField(userInput.sourceField);
+  const columnName = sanitizeMatColumnName(userInput.columnName);
+  return {
+    datatype: userInput.datatype,
+    sourceField,
+    columnName,
+  };
+}
+
+function sanitizeMatColumnSourceField(userInput: string) {
+  if (!/^[a-zA-Z0-9 _-]*$/.test(userInput)) {
+    throw new Error(
+      "Invalid input. Source field must only use alphanumeric characters, ' ', '_', or '-'"
+    );
+  }
+  return userInput;
+}
+
+function sanitizeMatColumnName(userInput: string) {
+  if (!/^[a-zA-Z_][a-zA-Z0-9_]*$/.test(userInput)) {
+    throw new Error(
+      "Invalid input. Column names must start with a letter or underscore and only use alphanumeric characters or '_'"
+    );
+  }
```

Comment:
We need a block list of reserved column names so we don't accidentally overwrite one of the default columns.

We also might want to block known SQL keywords (`count`, `select`, `from`, `where`, etc.) since it can cause some weirdness when querying.